pkg/challenge/response.go

```diff
@@ -29,23 +30,47 @@ func (o *OuterResponse) Open(privateEncryptionKey [32]byte) (*InnerResponse, err
 		return nil, fmt.Errorf("unmarshaling inner box: %w", err)
 	}
 
-	counterPartyPubKey, err := echelper.PublicPemToEcdsaKey(innerResponse.PublicSigningKey)
-	if err != nil {
-		return nil, fmt.Errorf("unmarshalling public ecdsa signing key from pem: %w", err)
+	if err := verifyWithKeyBytes(innerResponse.PublicSigningKey, innerResponseBytes, o.Sig); err != nil {
+		return nil, fmt.Errorf("verifying challenge signature: %w", err)
+	}
+
+	// neither signature 2 nor public signing key 2 was provided, return what we have
+	if (o.Sig2 == nil || len(o.Sig2) <= 0) && (innerResponse.PublicSigningKey2 == nil || len(innerResponse.PublicSigningKey2) <= 0) {
+		return &innerResponse, nil
+	}
+
+	// public key 2, no signature 2
+	if o.Sig2 == nil || len(o.Sig2) <= 0 {
+		return nil, fmt.Errorf("have public siging key 2 but no signature 2")
+	}
+
+	// signature 2, no public key 2
+	if innerResponse.PublicSigningKey2 == nil || len(innerResponse.PublicSigningKey2) <= 0 {
+		return nil, fmt.Errorf("have signature 2 but no public signing key 2")
 	}
 
-	if err := echelper.VerifySignature(*counterPartyPubKey, innerResponseBytes, o.Sig); err != nil {
-		return nil, fmt.Errorf("verifying challenge: %w", err)
+	if err := verifyWithKeyBytes(innerResponse.PublicSigningKey2, innerResponseBytes, o.Sig2); err != nil {
+		return nil, fmt.Errorf("verifying challenge signature 2: %w", err)
 	}
 
 	return &innerResponse, nil
 }
 
+func verifyWithKeyBytes(keyBytes []byte, msg []byte, sig []byte) error {
+	key, err := echelper.PublicPemToEcdsaKey(keyBytes)
```

Comment:
I like this. As we move from pem to der I think it's going to be easy to add a `strings.HasPrefix` here

Reply:
updated this to check for pem via HasPrefix else use der